api/datastore/sql/sql.go

```diff
@@ -667,6 +881,48 @@ func (ds *SQLStore) GetRoutesByApp(ctx context.Context, appID string, filter *mo
 	return res, nil
 }
 
+func (ds *SQLStore) GetFnByID(ctx context.Context, fnID string) (*models.Fn, error) {
```

Comment:
open ended question: it may be wise to require an app id as well as an fn id here to avoid security bugs where a user may be able to access another user's function which they shouldn't have access to by supplying the correct id. this would become an issue from some auth middleware, presumably, but would manifest as an issue at this level (because at auth time you wouldn't know the function doesn't exist on a certain other app until you look it up). 

for example:

* Bob has an app `poodles`, with function id `4242`. 
* Julie has an app `corgis` with a few functions on it. 
* Bob is Julie's sworn nemesis (ok, maybe not relevant). 
* Julie calls the api at `/v2/apps/corgis/fns/4242`, is authenticated against the `corgis` app correctly, and grabs Bob's function `4242` out of the db. Julie then sheathes Bob's beautiful poodle Fifi in the thick of Nordic winter (and maybe adopts it so that this is more of a rescue than a tale of malevolence?).

in any event, I have seen this before! maybe, we'd like to guard against it differently, but it's an issue at some level no less and the cost of adding app id to each of these endpoints is relatively minimal, as its already given in the API and we'd just be passing it down -- maybe, there are better ways though.

Reply:
My view is that access control is orthogonal to the API structure -  makes more sense (in my per-haps 90s oriented API design mentality) to solve authz on a resource-level properly rather than relying on API to help you here . 

Of course - getting the access control right (when you need access to the resource to assert on policy) is a harder problem that I leave as an excercise to the implementor of that feature. 

All of this is consolidating at the Data store level right now which is good as it's sound w.r.t most ACL decisions  but bad as it adds more complexity there. 

I /think/ you can plausibly implement a reasonably power Authz policy based on the data store abstraction by making decision after the subordinate call is complete but before the interceptor returns - this is what I've seen in the past (i.e. you can't make an ACL on a object till you've loaded it)